rectools/dataset/dataset.py

```diff
@@ -245,3 +258,72 @@ def get_raw_interactions(self, include_weight: bool = True, include_datetime: bo
         pd.DataFrame
         """
         return self.interactions.to_external(self.user_id_map, self.item_id_map, include_weight, include_datetime)
+
+    def construct_new_datasets(
```

Comment:
I'd call it `build_new_dataset_on_id_map` or smth like it. But we need to think about the conception first.

Right now it's not working properly in the conceptions that we use. We have a rule that all hot users (users that have interactions) always go before warm users (users that have only features). And we use this assumption in some of our impertant internal logic.

New method can break this rule because new users with interactions will have ids after warm users from previous dataset state. So we really can't make this a public method in the library.

Options are:
- Rewrite logic. We need to keep only hot users ids from previous state. Then add new hot users ids. Then add new warm user ids. I suppose we are dropping all previous warm users because we are totally changing the actual data. (All the same is for items) In this case I'd call it `update_with_new_data` or `rebuild_with_new_data` or `construct_with_new_data`
- Just don't make it a public method. Move it to the tests if we need it there. But then again write a correct public method for required logic. Also tests will be needed for it anyway

Reply:
Thanks for pointing it out. I wasn't aware of the warm users' assumption. And yeah, I expected this to be a controversial change. And that's why I haven't written the test yet 😅 

Originally, I expected for RecTools' `Dataset` to do similar to LightFM's `Dataset.fit_partial()`, but I found that RecTools `Dataset` is frozen.

I'm okay to make it private, or I can rebuild `IdMap` by concatenating new and old users/items something like:

```py
old_hot_user_id_map = IdMap.from_dict({e: i for e,i in  zip(ds.user_id_map.convert_to_external(ds.get_hot_users()), ds.get_hot_users())})
old_hot_item_id_map = ...

new_user_id = old_hot_user_id_map.add_ids(...
```

I'm not confident if it's okay to drop the previous warm user yet though,.

Reply:
Hmm, I attempt to drop old warm ids, but I face `UnknownIdError`.
https://github.com/MobileTeleSystems/RecTools/blob/f73d0541d6bef7291d149311f7be44dfb069ff7a/rectools/dataset/features.py#L121

Maybe, storing hot/warm marks like `hot_ids` in IdMap can be an option like `np.array([1, 0, 1])` where `1` represents hot and `0` represents warm id. Then, we can fetch hot or warm users/items explicitly.

Reply:
Okay, let me revive my dropping code and see what it happens on CI. I'll share it later.

Reply:
@blondered I think I completed implementing the "Rewrite logic" option, but let me know if you find anything else.